File: CarHomeWidget/src/main/java/com/anod/car/home/drawer/NavigationDrawer.java
package com.anod.car.home.drawer;

import com.anod.car.home.R;

import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import butterknife.ButterKnife;
import butterknife.Bind;

/**
 * @author alex
 * @date 2014-10-20
 */
public class NavigationDrawer {

    private final ActionBarDrawerToggle mDrawerToggle;

    private final CharSequence mTitle;

    private final CharSequence mDrawerTitle;

    private final NavigationAdapter mAdapter;

    @Bind(R.id.drawer_layout)
    DrawerLayout mDrawerLayout;

    @Bind(R.id.left_drawer)
    ListView mDrawerList;

    public NavigationDrawer(final AppCompatActivity activity, int appWidgetId) {
        ButterKnife.bind(this, activity);

        mTitle = mDrawerTitle = activity.getTitle();

        NavigationList items = new NavigationList(activity, appWidgetId);
        //items.addDefaults();

        mAdapter = new NavigationAdapter(activity, items);
        mDrawerList.setAdapter(mAdapter);
        mDrawerList.setOnItemClickListener(new DrawerItemClickListener());

        mDrawerToggle = new ActionBarDrawerToggle(activity, mDrawerLayout, R.string.drawer_open,
                R.string.drawer_close) {

            // Called when a drawer has settled in a completely closed state.
            public void onDrawerClosed(View view) {
                super.onDrawerClosed(view);
                activity.getSupportActionBar().setTitle(mTitle);
                activity.invalidateOptionsMenu(); // creates call to onPrepareOptionsMenu()
            }

            // Called when a drawer has settled in a completely open state.
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
                activity.getSupportActionBar().setTitle(mDrawerTitle);
                activity.invalidateOptionsMenu(); // creates call to onPrepareOptionsMenu()
            }
        };

        // Set the drawer toggle as the DrawerListener
        mDrawerLayout.setDrawerListener(mDrawerToggle);

        activity.getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        activity.getSupportActionBar().setHomeButtonEnabled(true);
    }

    public void syncState() {
        mDrawerToggle.syncState();
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        return mDrawerToggle.onOptionsItemSelected(item);
    }

    public void setSelected(int navigationItem) {
        mAdapter.setSelected(navigationItem);
    }

    public void refresh() {
        mAdapter.refresh();
    }

    private class DrawerItemClickListener implements ListView.OnItemClickListener {

        @Override
        public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
            selectItem(position);
        }
    }

    /** Swaps fragments in the main content view */
    private void selectItem(int position) {
        if (mAdapter.onClick(position)) {
            // Highlight the selected item, update the title, and close the drawer
            mDrawerList.setItemChecked(position, true);
            mDrawerLayout.closeDrawer(mDrawerList);
        }
    }
}

File: CarHomeWidget/src/main/java/com/anod/car/home/drawer/NavigationDrawer.kt
package com.anod.car.home.drawer

import com.google.android.material.navigation.NavigationView
import androidx.drawerlayout.widget.DrawerLayout
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import android.view.MenuItem
import android.view.View

import com.anod.car.home.R

/**
 * @author alex
 * @date 2014-10-20
 */
class NavigationDrawer(activity: AppCompatActivity, appWidgetId: Int) {

    private val drawerToggle: ActionBarDrawerToggle

    private val title = activity.title

    private val drawerTitle = activity.title

    private val drawerLayout: androidx.drawerlayout.widget.DrawerLayout = activity.findViewById(R.id.drawer_layout)

    private val drawerView: NavigationView? = activity.findViewById(R.id.left_drawer)

    private val selection: NavigationDrawerSelection = NavigationDrawerSelection(activity, appWidgetId)

    init {
        if (appWidgetId == 0) {
            drawerView?.menu?.findItem(R.id.nav_current_widget_group)?.isVisible = false
        }

        drawerView?.setNavigationItemSelectedListener { menuItem ->
            // Add code here to update the UI based on the item selected
            // For example, swap UI fragments here

            selection.onClick(menuItem.itemId)

            // close drawer when item is tapped
            drawerLayout.closeDrawers()

            true
        }

        drawerToggle = object : ActionBarDrawerToggle(activity, drawerLayout, R.string.drawer_open,
                R.string.drawer_close) {

            // Called when a drawer has settled in a completely closed state.
            override fun onDrawerClosed(view: View) {
                super.onDrawerClosed(view)
                activity.supportActionBar!!.title = title
                activity.invalidateOptionsMenu() // creates call to onPrepareOptionsMenu()
            }

            // Called when a drawer has settled in a completely open state.
            override fun onDrawerOpened(drawerView: View) {
                super.onDrawerOpened(drawerView)
                activity.supportActionBar!!.title = drawerTitle
                activity.invalidateOptionsMenu() // creates call to onPrepareOptionsMenu()
            }
        }

        // Set the drawer toggle as the DrawerListener
        drawerLayout.addDrawerListener(drawerToggle)

        activity.supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        activity.supportActionBar!!.setHomeButtonEnabled(true)
    }

    fun syncState() {
        drawerToggle.syncState()
    }

    fun onOptionsItemSelected(item: MenuItem): Boolean {
        return drawerToggle.onOptionsItemSelected(item)
    }

    fun setSelected(itemId: Int) {
        drawerView?.menu?.findItem(itemId)?.isChecked = true
    }

}
